cmd/juju/application/deployer/deployer.go

```diff
@@ -54,25 +75,202 @@ func NewDeployerFactory(dep DeployerDependencies) DeployerFactory {
+	// TODO (cderici): check the validity of the comments belowe
+	// NOTE: Here we select the series using the algorithm defined by
+	// `seriesSelector.charmSeries`. This serves to override the algorithm found
+	// in `charmrepo.NewCharmAtPath` which is outdated (but must still be
+	// called since the code is coupled with path interpretation logic which
+	// cannot easily be factored out).
+
+	// NOTE: Reading the charm here is only meant to aid in inferring the
+	// correct series, if this fails we fall back to the argument series. If
+	// reading the charm fails here it will also fail below (the charm is read
+	// again below) where it is handled properly. This is just an expedient to
+	// get the correct series. A proper refactoring of the charmrepo package is
+	// needed for a more elegant fix.
+	var (
+		imageStream string
+		seriesName  string
+	)
+	if !d.base.Empty() {
+		var err error
+		seriesName, err = series.GetSeriesFromBase(d.base)
+		if err != nil {
+			return "", "", errors.Trace(err)
+		}
+	}
+
+	ch, err := d.charmReader.ReadCharm(charmOrBundle)
+	if err == nil {
+		modelCfg, err := getModelConfig(getter)
+		if err != nil {
+			return "", "", errors.Trace(err)
+		}
+
+		imageStream = modelCfg.ImageStream()
+		workloadSeries, err := SupportedJujuSeries(d.clock.Now(), seriesName, imageStream)
+		if err != nil {
+			return "", "", errors.Trace(err)
+		}
+
+		supportedSeries, err := corecharm.ComputedSeries(ch)
+		if err != nil {
+			return "", "", errors.Trace(err)
+		}
+		seriesSelector := seriesSelector{
+			seriesFlag:          seriesName,
+			supportedSeries:     supportedSeries,
+			supportedJujuSeries: workloadSeries,
+			force:               d.force,
+			conf:                modelCfg,
+			fromBundle:          false,
+		}
+
+		if len(supportedSeries) == 0 {
+			logger.Warningf("%s does not declare supported series in metadata.yml", ch.Meta().Name)
+		}
+
+		seriesName, err = seriesSelector.charmSeries()
+		if err = charmValidationError(ch.Meta().Name, errors.Trace(err)); err != nil {
+			return "", "", errors.Trace(err)
+		}
+	}
+	return seriesName, imageStream, nil
+}
+
 // GetDeployer returns the correct deployer to use based on the cfg provided.
 // A ModelConfigGetter needed to find the deployer.
 func (d *factory) GetDeployer(cfg DeployerConfig, getter ModelConfigGetter, resolver Resolver) (Deployer, error) {
 	d.setConfig(cfg)
-	maybeDeployers := []func() (Deployer, error){
-		d.maybeReadLocalBundle,
-		func() (Deployer, error) { return d.maybeReadLocalCharm(getter) },
-		d.maybePredeployedLocalCharm,
-		func() (Deployer, error) { return d.maybeReadRepositoryBundle(resolver) },
-		d.repositoryCharm, // This always returns a Deployer
-	}
-	for _, d := range maybeDeployers {
-		if deploy, err := d(); err != nil {
+
+	// Determine the type of deploy we have
+	var dk DeployerKind
+
+	userCharmURL, err := resolveCharmURL(d.charmOrBundle, d.defaultCharmSchema)
```

Comment:
this will fail on a local charm, nor is it used until ln 221. move closer to where its used.